docs/madr/decisions/020-kds-redesign.md

```diff
@@ -0,0 +1,247 @@
+# KDS Redesign
+
+* Status: accepted
+
+Technical Story: https://github.com/kumahq/kuma/issues/4354
+
+## Context and Problem Statement
+
+The current implementation of KDS is quite complex, and there may be some bugs, such as the one found at https://github.com/kumahq/kuma/pull/5373, which was very difficult to identify and resolve. The use of a bi-directional stream for configuration exchange between the zone and the global can also be challenging to implement. Additionally, KDS operates as a State of The World, meaning that whenever an item in the resource type changes, the entire state is sent. While this doesn't appear to be a problem when synchronizing data from Global to Zone, it may be less efficient when there are many Dataplane changes. There are several improvements we want to make in the future such as:
+* working with different older versions of the control plane on the zone,
+* creating resources on the zone and syncing them to the global,
+* creating separate gRPC services for synchronization between the zone and global,
+* supporting multiple control plane tenants,
+* prioritizing sync of operational over informational resources,
+* changing `default` namespace for sync to global to system one (`kuma-system`), 
```

Comment:
What's the impact of this when global is on k8s? I guess we'll need to think about the upgrade path a little?

Reply:
I feel like there might be an issue with duplicated Dataplanes in the insight.

When we upgrade global:
1. Global starts and saves data in `kuma-system` namespace
2. Process responsible for the generation of DataplaneInsights, list Dataplanes from all the namespaces
3. DataplaneInsight has an overview with duplicated dataplanes

We could introduce a namespace parameter to read dataplanes from the specific namespace. Namespace is defined on the global level so it might don't need changes on the zone.

Reply:
Yes maybe the meshInsight generator needs to be aware of which version the zone is using to be able to pick the right namespace for a specific zone (or at least ignore the ones in the default namespace)

Reply:
> think about the upgrade path a little?

A lot

> What's the impact of this when global is on k8s?

The impact is also on Zone CPs. There are namespace scoped resources synced from global to zones

Reply:
On zone CPs we are storing resources in `default` namespace or `kuma-system` in case of new policies. I see that in kds we don't send information from zone to global or global to zone about where to store the resources. We do our assumptions based on the resource type.

Reply:
So I've done some testing it looks like 'sync' check the current state of the resources and new state of the resources. If in the incoming state, there is no resource in this namespace, it removes it. E.g.:global is running with an insert to `default` namespace and have `zoneIngress` resource in the namespace `default`, we are upgrading global to the version with the insertion to the namespace `kuma-system`, resources are synced and the `zoneIngress` resource in `default` namespace is removed while `kuma-system` is created. There should be no cleanup necessary because there won't be duplicate.